### PR TITLE
Desugar simple UCS and add old tests to the staging area

### PR DESCRIPTION
**Pull request by chengluyu**
This PR adds `Desugarer` to hkmc2. It supports simple UCS expressions now.

### CODE REVIEW

File: hkmc2/shared/src/main/scala/hkmc2/bbml/TypeSimplifier.scala [hunk 1]
```diff
@@ -5,7 +5,7 @@ import scala.collection.immutable.{SortedMap, SortedSet}
 import scala.util.chaining._
 
 import mlscript.utils.*, shorthands.*
-import utils.TraceLogger
+import hkmc2.utils.TraceLogger
```
Comment (LPTK) [inline hkmc2/shared/src/main/scala/hkmc2/bbml/TypeSimplifier.scala]:
```suggestion
import hkmc2.utils.*
```

File: hkmc2/shared/src/main/scala/hkmc2/syntax/ParseRule.scala [hunk 2]
```diff
@@ -167,14 +169,35 @@ object ParseRule:
         }
     ,
     Kw(`if`):
-      ParseRule("`if` keyword")(
+      ParseRule("`if`")(
```
Comment (LPTK) [inline hkmc2/shared/src/main/scala/hkmc2/syntax/ParseRule.scala]: Why this change?

File: hkmc2/shared/src/main/scala/hkmc2/syntax/ParseRule.scala [hunk 3]
```diff
@@ -167,14 +169,35 @@ object ParseRule:
...
         Expr(
-          ParseRule("`if` expression"):
+          ParseRule("`if` split")(
```
Comment (LPTK) [inline hkmc2/shared/src/main/scala/hkmc2/syntax/ParseRule.scala]: This is for error messages. Why introduce internal jargon (if "split") here?

File: hkmc2/shared/src/main/scala/hkmc2/syntax/Tree.scala [hunk 4]
```diff
@@ -73,13 +82,47 @@ enum Tree extends AutoLocated:
...
+    case DecLit(value) => value.toString
+    case StrLit(value) => value.escaped
+    case UnitLit(value) => if value then "undefined" else "null"
+    case BoolLit(value) => value.toString
+    case Block(stmts) => stmts.map(_.describe).mkString("{ ", "; ", " }")
+    case Let(lhs, rhs, body) => s"let $lhs = $rhs in $body"
+    case TermDef(k, symName, alphaName, rhs) =>
+      val name = symName.orElse(alphaName).fold("")(n => s" ${n.describe}")
+      val sign = rhs.fold("")(s => s" : ${s.describe}")
+      s"${k.str}$name$sign"
```
Comment (LPTK) [inline hkmc2/shared/src/main/scala/hkmc2/syntax/Tree.scala]: This is crazy. I assume this is Copilot of ChatGPT junk. The goal of `describe` is simply to name what tree we're looking at using a word or two.

Comment (chengluyu) [reply hkmc2/shared/src/main/scala/hkmc2/syntax/Tree.scala]: You're absolutely right. I didn't look into this case. I will remove it soon.

File: hkmc2/shared/src/main/scala/hkmc2/semantics/Desugarer.scala [hunk 5]
```diff
@@ -5,31 +5,32 @@ import syntax.{Keyword, Tree}, Tree.*
 import mlscript.utils.*, shorthands.*
 import Message.MessageContext
 import utils.TraceLogger
+import scala.annotation.showAsInfix
```
Comment (LPTK) [inline hkmc2/shared/src/main/scala/hkmc2/semantics/Desugarer.scala]: Unused import...

File: hkmc2/shared/src/test/mlscript/ucs/syntax/SimpleUCS.mls [hunk 6]
```diff
@@ -93,20 +84,14 @@ fun f(x, y, z) =
       B then "bruh"
     A then "oui"
 
-:fixme
-// :e
+
+// :e // Note: should report missing final else clause.
```
Comment (LPTK) [inline hkmc2/shared/src/test/mlscript/ucs/syntax/SimpleUCS.mls]: I guess this should be both `:e` and `:fixme`, which should not be reported as a test failure.

File: hkmc2/shared/src/test/mlscript/ucs/syntax/Split.mls [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+:todo
+if
+  x
+  is
+    A then "A"
+//│ /!!!\ Uncaught error: scala.NotImplementedError: List()
+
+x
+  + 1
```
Comment (LPTK) [inline hkmc2/shared/src/test/mlscript/ucs/syntax/Split.mls:105]: This is misparsed. Clearly it should not parse the same as `x({ +(1) })`. You can see `x` is not being applied and `+`  does not have a single argument...

File: hkmc2/shared/src/main/scala/hkmc2/semantics/Elaborator.scala [hunk 8]
```diff
@@ -280,8 +339,13 @@ class Elaborator(tl: TraceLogger)(using raise: Raise, state: State):
             (name, tas, S(ps), newCtx2)
           case id: Ident =>
             (id, Nil, N, ctx)
+          case InfixApp(lhs, Keyword.`:`, rhs) =>
+            // For self-annotations
```
Comment (LPTK) [inline hkmc2/shared/src/main/scala/hkmc2/semantics/Elaborator.scala]: What are self-annotations?

File: hkmc2/shared/src/main/scala/hkmc2/semantics/Split.scala [hunk 9]
```diff
@@ -3,70 +3,103 @@ package semantics
 
 import mlscript.utils.*, shorthands.*
 import syntax.*
+import scala.annotation.tailrec
```
Comment (LPTK) [inline hkmc2/shared/src/main/scala/hkmc2/semantics/Split.scala]: Unused import

File: hkmc2/shared/src/main/scala/hkmc2/syntax/ParseRule.scala [hunk 10]
```diff
@@ -249,6 +306,19 @@ object ParseRule:
...
+    //       End(N),
+    //       Kw(`else`):
+    //         ParseRule(s"`then` operator `else` clause")(
+    //           Expr(ParseRule(s"`then` operator `else` body")(End(()))):
+    //             case (body, _) => S(body)
+    //         )
+    //     )):
+    //       case (consequent, N) => (test: Tree) => InfixApp(test, `then`, consequent)
+    //       case (consequent, S(default)) => (test: Tree) =>
+    //         InfixApp(test, `then`, InfixApp(consequent, `else`, default)),
```
Comment (LPTK) [inline hkmc2/shared/src/main/scala/hkmc2/syntax/ParseRule.scala]: Is this commented code still relevant? Or should it be removed?

**Review by LPTK — CHANGES_REQUESTED** [review]
Here are some comments. I'll still merge the PR to avoid blocking progress, but I'll leave the comments open and you should make sure to address them later on, when you have time. I'll make the merge after you address the serious problem I told you about.

File: hkmc2/shared/src/test/mlscript/ucs/syntax/SimpleUCS.mls (hunk 6 above)
Comment (chengluyu) [reply hkmc2/shared/src/test/mlscript/ucs/syntax/SimpleUCS.mls]: It is reported unless we disable the following lines.

https://github.com/LPTK/mlscript/blob/20fa71a262dcf659993172f9c6cfaafb998b77a6/hkmc2/jvm/src/test/scala/hkmc2/DiffMaker.scala#L167-L169

File: hkmc2/shared/src/main/scala/hkmc2/syntax/Parser.scala [hunk 11]
```diff
@@ -547,27 +548,33 @@ abstract class Parser(
...
-      consume
-      val default = expr(0, true)
-      Modified(kw, default) :: splitContOf()
-    case (tok @ IDENT(opStr, true), loc) :: _ if opPrec(opStr)._1 > 0 =>
-      consume
-      App(Ident(opStr).withLoc(S(loc)), expr(0, false)) :: splitContOf()
-    case (_, loc) :: _ =>
-      err(msg"Expect an operator" -> S(loc) :: Nil)
-      errExpr :: Nil
+  /** Parse a operator block. Each block item should be a binary operator
```
Comment (LPTK) [inline hkmc2/shared/src/main/scala/hkmc2/syntax/Parser.scala]:
```suggestion
  /** Parse an operator block. Each block item should be a binary operator
```

**Review by LPTK — APPROVED** [review]
